core/src/main/java/org/apache/brooklyn/core/mgmt/internal/ManagementNodeStateListenerManager.java

```diff
@@ -0,0 +1,175 @@
+ *      for how we get notified of the state-change.
+ */
+public class ManagementNodeStateListenerManager implements ManagementNodeStateListener {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ManagementNodeStateListenerManager.class);
+
+    private final ManagementContextInternal mgmt;
+    
+    private final Object mutex = new Object();
+
+    private final List<ManagementNodeStateListener> listeners = Lists.newCopyOnWriteArrayList();
+    private ManagementNodeState lastPublishedVal;
+    
+    private final AtomicInteger listenerQueueSize = new AtomicInteger();
+    
+    private ListeningExecutorService listenerExecutor = MoreExecutors.listeningDecorator(Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+            .setNameFormat("brooklyn-managementnodestate-listener-%d")
+            .build()));
+
+    public ManagementNodeStateListenerManager(ManagementContextInternal managementContext) {
+        this.mgmt = checkNotNull(managementContext, "managementContext");
+        
+        // Register a coercion from String->ManagementNodeStateListener, so that MANAGEMENT_NODE_STATE_LISTENERS defined in brooklyn.cfg
+        // will be instantiated, given their class names.
+        TypeCoercions.BrooklynCommonAdaptorTypeCoercions.registerInstanceForClassnameAdapter(
+                new ClassLoaderUtils(this.getClass(), managementContext), 
+                ManagementNodeStateListener.class);
+
+        // Although changing listeners to Collection<ManagementNodeStateListener> is valid at compile time
+        // the collection will contain any objects that could not be coerced by the function
+        // declared above. Generally this means any string declared in brooklyn.properties
+        // that is not a ManagementNodeStateListener.
+        Collection<?> rawListeners = managementContext.getBrooklynProperties().getConfig(BrooklynServerConfig.MANAGEMENT_NODE_STATE_LISTENERS);
+        if (rawListeners != null) {
+            for (Object obj : rawListeners) {
+                if (obj == null) {
+                    throw new NullPointerException("null listener in config " + BrooklynServerConfig.MANAGEMENT_NODE_STATE_LISTENERS);
+                } else if (!(obj instanceof ManagementNodeStateListener)) {
+                    throw new ClassCastException("Configured object is not a "+ManagementNodeStateListener.class.getSimpleName()+". This probably means coercion failed: " + obj);
+                } else {
+                    ManagementNodeStateListener listener = (ManagementNodeStateListener) obj;
+                    if (listener instanceof ManagementContextInjectable) {
+                        ((ManagementContextInjectable) listener).setManagementContext(managementContext);
+                    }
+                    listeners.add((ManagementNodeStateListener)listener);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void onStateChange(ManagementNodeState state) {
```

Comment:
seems a bit strange that we can call onStateChange with the same state and have the ManagementNodeStateListenerManager filter it out. Feel like it should be the responsibility of the HighAvailabilityManagerImpl, but it is neat to include it in ManagementNodeStateListenerManager

Reply:
Agreed it's a bit strange. But I didn't want to complicate the `HighAvailabilityManagerImpl` more. It would only be a few lines in there (a new field to store the last value, etc) - I could move it there if you think it's better @Graeme-Miller?

Reply:
not sure, but I think it is fine as it is. Your call.